### PR TITLE
Allow user-provided User-Agent request header

### PR DESCRIPTION
**Pull request by polRk**
#1271

### CODE REVIEW

File: packages/connect/src/protocol-grpc/request-header.ts [hunk 1]
```diff
@@ -32,18 +32,16 @@ export function requestHeader(
...
 ): Headers {
-  const result = new Headers(userProvidedHeaders ?? {});
+  const result = new Headers(userProvidedHeaders);
   result.set(
     headerContentType,
     useBinaryFormat ? contentTypeProto : contentTypeJson,
   );
-  // Note that we do not strictly comply with gRPC user agents.
-  // We use "connect-es/1.2.3" where gRPC would use "grpc-es/1.2.3".
-  // See https://github.com/grpc/grpc/blob/c462bb8d485fc1434ecfae438823ca8d14cf3154/doc/PROTOCOL-HTTP2.md#user-agents
```
Comment (timostamm) [inline packages/connect/src/protocol-grpc/request-header.ts:42]: If you follow the link in the comment, you'll see that the gRPC spec recommends that clients provide a user-agent.

Comment (polRk) [reply packages/connect/src/protocol-grpc/request-header.ts:42]: Yes, and I want to pass my user-agent: to understand who my consumers are, add the runtime version, the library version

Comment (polRk) [reply packages/connect/src/protocol-grpc/request-header.ts:42]: @timostamm I've updated PR

**Review by timostamm — CHANGES_REQUESTED** [review]
Thanks for the PR. If you can update it to only make the change to not set the header if the user already provided one, we can merge it.

I understand the idea, but we do not want to make a breaking change to user agent strings. Please avoid unnecessary changes to whitespace (`npx turbo run format` runs prettier).

**Comment by polRk** [comment]
> Thanks for the PR. If you can update it to only make the change to not set the header if the user already provided one, we can merge it.

So, are you saying that the logic for generating the grpc-js-* string should be removed, and we should just pass the header instead?

**Comment by timostamm** [comment]
> So, are you saying that the logic for generating the grpc-js-* string should be removed

Yes.

> and we should just pass the header instead?

Currently, we're always setting the `User-Agent` header, regardless of whether the user has already provided one. For example:

```ts
await client.say(
  { sentence: "hello" },
  {
    headers: {
      "User-Agent": "foo",
    }
  }
);
```

The `User-Agent: foo` is overwritten.

Following https://github.com/connectrpc/connect-es/issues/1271#issuecomment-2411048990, we should change the behavior, and not overwrite the user-provided value, but continue to set our value if the user didn't provide one.

The important part is that we do it consistently, not just in the gRPC transport, but also for gRPC-Web and the Connect transport.

Adding a transport option for a default user-agent is a nice idea, but it raises the question whether we shouldn't have an option for default headers instead. For now, the best action is to not add this option, and consider this change separately.

**Comment by polRk** [comment]
> Currently, we're always setting the `User-Agent` header, regardless of whether the user has already provided one. For example:

The problem is:

1. Passing headers would require changing the code throughout the entire codebase. In my previous project, I had thousands of gRPC requests, and modifying all of them would be difficult.

2. Using interceptors introduces unnecessary overhead, especially when using AsyncLocalStorage.

**Comment by timostamm** [comment]
To fix #1271, we can allow to pass a custom user agent via `CallOptions.headers`, which is a reasonable incremental change.

We can discuss other features separately.

**Comment by polRk** [comment]
@timostamm

File: packages/connect/src/protocol-grpc/request-header.spec.ts [hunk 2]
```diff
@@ -34,7 +34,7 @@ describe("requestHeader", () => {
       "user-agent",
     ]);
     expect(headers.get("Content-Type")).toBe("application/grpc+proto");
-    expect(headers.get("User-Agent")).toMatch(/^connect-es\/\d+\.\d+\.\d+/);
+    expect(headers.get("User-Agent")).toMatch(/^connect-es\/.*/);
```
Comment (timostamm) [inline packages/connect/src/protocol-grpc/request-header.spec.ts:37]: Let's not modify this test. It makes sure the default user agent doesn't regress.

File: packages/connect/src/protocol-grpc/request-header.spec.ts [hunk 3]
```diff
@@ -48,6 +48,17 @@ describe("requestHeader", () => {
     expect(headers.get("Grpc-Timeout")).toBe("10m");
   });
 
+  it("should create request headers with userAgent", () => {
```
Comment (timostamm) [inline packages/connect/src/protocol-grpc/request-header.spec.ts:51]: Thanks for the test!

File: packages/connect/src/protocol-grpc/request-header.ts [hunk 4]
```diff
@@ -37,13 +37,18 @@ export function requestHeader(
...
-  result.set(headerUserAgent, "CONNECT_ES_USER_AGENT");
+
+  if (!result.has(headerUserAgent)) {
```
Comment (timostamm) [inline packages/connect/src/protocol-grpc/request-header.ts:41]: Nice and simple 👍

**Review by timostamm — COMMENTED** [review]
Looks great. 

Can you apply the same change to the Connect and gRPC-Web transports for consistency? It would be very confusing if this only works with gRPC. The transports from `@connectrpc/connect-web` don't set a user-agent, no need to update them.

**Comment by polRk** [comment]
@timostamm Greetings! Will you be able to watch it?

**Pull request merged**